### PR TITLE
Add support for PCI device pass through

### PR DESCRIPTION
**Pull request by romantomjak**
This PR introduces the `pci_devices` block to allow mapping host's PCI devices into the guest VM. This is often used to passthrough a graphics card or a network adapter. Devices that are mapped into a guest VM are no longer available on the host.

### CODE REVIEW

File: builder/proxmox/common/step_start_vm.go
Comment (romantomjak) [inline builder/proxmox/common/step_start_vm.go:363]: Didn't see any examples of optional values in the project, so came up with this. The use case is for setting a default bool value if none was given. Let me know if there is a better way to handle "unset" values 👍

File: docs/builders/iso.mdx
Comment (romantomjak) [inline docs/builders/iso.mdx:362]: This is a drive-by fix for what I hope will rectify the following situation on the website:

image

File: builder/proxmox/common/config_test.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config_test.go:385]: Regarding the host, do we have a way to validate if they're well formed? This would be a good check to add in the `Prepare` function for the config, this way we can ensure those are not malformed by accident before attempting to build and get an obscure error from `qemu`.

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:443]: Since `Trilean` exists, I believe this can be removed, unless there's something that `Trilean` doesn't support and you think it's still relevant? Always open to discuss :)

File: docs/builders/clone.mdx
Comment (lbajolet-hashicorp) [inline docs/builders/clone.mdx:260]: FYI: those options are compiled from the Go code documentation, into the partial configuration docs you've included in this PR. Those partials can be included in the documentation here, and in ISO.mdx, so you won't have to copy-paste those (which can easily diverge in the future when we modify them later).

If you want an example of this, @mabeett pushed a PR recently and did something similar in their code on PR #195: https://github.com/hashicorp/packer-plugin-proxmox/pull/195/files#diff-d395f7f43fac39fb86461a2b4239b646077ca6c2989e2fa3b0c6f174038b29ebR359

File: docs/builders/clone.mdx
Comment (lbajolet-hashicorp) [inline docs/builders/clone.mdx:239]: Same for this docs exerpt, and the following example, they can be extracted into the common config documentation in the Go code, this way that'll be automatically compiled into a partial which you can include here.

**Review by lbajolet-hashicorp — CHANGES_REQUESTED** [review]
Hi @romantomjak,

Thanks for this PR! I've taken a look at the options you're introducing and responded to your comments. I also added a few comments and suggestions that should be fixed before we can merge, but overall this looks good to me.

I'll let you address those comments, and I'll circle back later.
As always please let me know if you need any kind of help!

File: docs/builders/iso.mdx
Comment (lbajolet-hashicorp) [reply docs/builders/iso.mdx:362]: FYI, PR #200 fixes this problem, so I'll merge this and you can leave this change out of yours after that!

Comment (romantomjak) [reply docs/builders/iso.mdx:362]: Thank you, that's really useful! 🎉

File: builder/proxmox/common/config_test.go
Comment (romantomjak) [reply builder/proxmox/common/config_test.go:385]: Yeah, that's a good shout! I'll see if I can bottom it out 👍

File: builder/proxmox/common/step_start_vm.go
Comment (romantomjak) [reply builder/proxmox/common/step_start_vm.go:363]: Thanks, these are excellent suggestions! I really like the idea of just inverting the flag 💯

File: builder/proxmox/common/config_test.go
Comment (romantomjak) [reply builder/proxmox/common/config_test.go:472]: I found a regex in proxmox API viewer that we can use 👇 
image

File: builder/proxmox/common/step_start_vm_test.go
Comment (romantomjak) [inline builder/proxmox/common/step_start_vm_test.go:446]: Added this test to make sure the value is actually inverted when the config is assembled

File: docs/builders/clone.mdx
Comment (romantomjak) [inline docs/builders/clone.mdx:368]: De-dented header size. The config reference section has h2 headers, so I think this must be h3.

File: docs/builders/clone.mdx
Comment (romantomjak) [reply docs/builders/clone.mdx]: Done ✅

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:208]: I wonder given how the comments are phrased, are configurations with either `host` or `mapping` accepted? My understanding with `required` is that both options will be required if we note them as `required:"true"`.
In cases like these, I would suggest leaving `required` as false, and checking that one or the other is set, and not both.

Comment (lbajolet-hashicorp) [reply builder/proxmox/common/config.go:208]: I believe that your tests won't catch this since it's related to parsing configurations, this may be tested via acceptance tests, but AFAIK we don't have those setup for proxmox (we'd need some infrastructure for this).
In the meantime, could you manually check the point I bring here? I would like to make sure this won't be an issue before merging this PR.

File: docs/builders/iso.mdx
Comment (lbajolet-hashicorp) [inline docs/builders/iso.mdx:37]: I may misunderstand something, but this section points to the external resources, so this feels a bit out of place imho. I would suggest leaving it as-is since it points out to the communicator docs outside of the plugin, while the PCI Devices/VirtIO Rng are sections within this plugin.

File: builder/proxmox/common/config_test.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config_test.go:452]: Would it be possible to add a test with both `Mapping` and `Host` set to make sure `Prepare` returns an error in this case?

**Review by lbajolet-hashicorp — COMMENTED** [review]
Hi @romantomjak,

Thanks for the quick reroll, this looks good to me already :)

I left a couple more comments on the reworked PR, but as I said above, this looks really good, so I don't think there's a lot left to change here.

That being said, I'll let you have a look at what I point out/suggest, and let me know which deserve some attention.

I'll circle back later when you've had time to address those.

Thanks again for this PR!

File: builder/proxmox/common/config.go
Comment (romantomjak) [reply builder/proxmox/common/config.go:208]: Looks like you're right about this - [both cannot be set](https://github.com/proxmox/qemu-server/blob/742da46a44f7d882620ae0d8d8cc08ab7b3f8dcc/PVE/QemuServer/PCI.pm#L428) 👍

File: docs/builders/iso.mdx
Comment (romantomjak) [reply docs/builders/iso.mdx:37]: Sorry, I don't quite follow. Are you saying to revert this change entirely or just keep the communicator in a seperate paragraph?

Comment (lbajolet-hashicorp) [reply docs/builders/iso.mdx:37]: Oh I meant only for the communicator here, since it is a different page I think it makes sense to keep a reference to it, it's just that mixing in anchors to sections of the current page seems out-of-place to me.

Comment (romantomjak) [reply docs/builders/iso.mdx:37]: Cool, yep that makes sense 👍

File: builder/proxmox/common/step_start_vm.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/step_start_vm.go:305]: As we discuss in #207, `setDeviceParamIfDefined` only behaves correctly when dealing with `string` values, and since these are boolean options I would suggest replacing those invocations by a `if`.

Comment (romantomjak) [reply builder/proxmox/common/step_start_vm.go:305]: Sure, but I'm not entirely convinced we should do it?

I don't think it's enough to just check if the proxmox default is false and then exclude the value if the bool is false. For example, a machine has set `legacy_igd=true`. You attempt to clone it, but decide against the flag so you set it to false in packer config in which case it would be just omitted and the result would not be what you expected. To me that is worse than perhaps somewhat redundantly displaying the default value in proxmox web interface. wdyt?

File: builder/proxmox/common/config.go
Comment (romantomjak) [reply builder/proxmox/common/config.go:184]: Sorry, that was vscode automatically inserting tabs there for some reason whenever I press ctrl+s 🤔

File: builder/proxmox/common/step_start_vm.go
Comment (romantomjak) [reply builder/proxmox/common/step_start_vm.go:305]: Just to clarify, this is what I mean when I say "displaying default values in proxmox".

The ROM-Bar field defaults to `true`, so in proxmox when you add a PCI device it is not visible in the "hardware list view" (the line that starts with `hostpci0` behind the modal) but because we set the value to it's default when creating the VM - it gets appeneded to the config and thus also visible in the interface (even though it's at default value)

image

Comment (lbajolet-hashicorp) [reply builder/proxmox/common/step_start_vm.go:305]: In this case I would say this is a matter of intent, and how we want to convey it. `setDeviceParamIfDefined` implies that we set the boolean only if it's true (the function misbehaves with such attributes, but this is an alternate discussion to have).
Your description in my opinion warrants getting rid of both the function call, and the `if` then, and we should always set their value to the boolean's (or the inverse in the case of `HideROMBAR`), and not within `setDeviceParamIfDefined`, if only to make it clear this is intended to be part of the payload, regardless if it's been set or not.
Does that make sense?

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:200]: Because of the tab here I would think this won't render as expected, I suggest this is placed on the same level as the beginning of the code block.

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [inline builder/proxmox/common/config.go:186]: In HCL2 generally, the contents of a block are intented (with 2 spaces) relative to their immediate parent, I suggest we do the same here, is this OK?

Comment (romantomjak) [reply builder/proxmox/common/config.go:186]: Sorry, I can't get my editor to do that. If I add two spaces it just reindents everything extremely weirdly [like the rng0](https://github.com/romantomjak/packer-plugin-proxmox/blob/edcd09a5834cb994c848f6145d68b4573a2ade10/builder/proxmox/common/config.go#L130-L138)

File: builder/proxmox/common/step_start_vm.go
Comment (romantomjak) [reply builder/proxmox/common/step_start_vm.go]: Yeah, I think so. Can you please take a look now?

File: builder/proxmox/common/config.go
Comment (lbajolet-hashicorp) [reply builder/proxmox/common/config.go:186]: It's OK! I can do this change in a subsequent PR either way, so I think I'll merge your PR if everything checks out, and fix this later :)

File: builder/proxmox/common/step_start_vm.go
Comment (lbajolet-hashicorp) [reply builder/proxmox/common/step_start_vm.go]: Looks good to me then!

**Review by lbajolet-hashicorp — APPROVED** [review]
With the last changes, LGTM!

I'll merge this as soon as tests are green. Many thanks once more @romantomjak!

**Pull request merged**
